### ISSUE TITLE
Funny behavior with lambda=0 and changes to random seed...

### ISSUE THREAD

**Issue by dataforager** [dmlc/xgboost#2656]
Hello,

   Just doing some more reproducibility checks and came across something odd that we wanted to get your feedback on.

   When we set 'stochastic' parameters (subsample, colsample_bylevel, etc.) to 1.0 (so using 'deterministic' xgboost), lambda to 0 and nthread to 1, we see different predicted probabilities (here using a 'binary:logistic' objective) on the same held-out test set with two different random seeds (random_state was set to 1 or 2).

   Funnily enough, we don't observe this pattern when lambda > 0.

   I should also mention we're using gbtree as the boosting method.

   That's why it seemed weird to us because lambda is the L2 regularization parameter for linear
boosting (correct?).

   So, we're wondering why the absence of L2 regularization seems to introduce randomness in gbtree results?

   We see this pattern for two different models and on two different datasets.

   Exact parameterizations of the two models tested appear below:

   MODEL 1 PARAMETERS:
    num.rounds <- 61
    params <- list(eta = 0.127453813933,
                   gamma = 0.0287061314255, 
                   max_depth = 16,
                   min_child_weight = 0,
                   max_delta_step = 0,
                   tree_method = 'approx',
                   eval_metric="auc",
                   subsample = 1,
                   colsample_bytree=1,
                   colsample_bylevel=1)

   MODEL 2 PARAMETERS:
    num.rounds <- 64
    params <- list(eta = 0.174279496229,
                   gamma = 0.0125680794894,
                   max_depth = 18,
                   min_child_weight = 0,
                   max_delta_step = 0,
                   tree_method = 'exact',
                   eval_metric="auc",
                   subsample = 1,
                   colsample_bytree=1,
                   colsample_bylevel=1)   

   Thanks a lot for your help!

## Environment info
Operating System: Ubuntu 16.04 LTS

Compiler: GNU gcc/g++

Package used (python/R/jvm/C++): R

`xgboost` version used: 0.6-4

R Session Info:
R version 3.4.1 (2017-06-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.3 LTS

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.6.0
LAPACK: /usr/lib/lapack/liblapack.so.3.6.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] xgboost_0.6-4

loaded via a namespace (and not attached):
[1] compiler_3.4.1    magrittr_1.5      Matrix_1.2-10     tools_3.4.1       stringi_1.1.5    
[6] grid_3.4.1        data.table_1.10.4 lattice_0.20-35

**Comment by Laurae2** [comment]
@dataforager `lambda` is usable with `gbtree`. In fact, all `gblinear` parameters are usable with `gbtree` (such as `alpha`, `lambda`, and `lambda_bias`).

I cant seem to reproduce the non-reproducibility issue with this code:

```r
library(xgboost)

data(agaricus.train, package='xgboost')
data(agaricus.test, package='xgboost')

dtrain <- xgb.DMatrix(agaricus.train$data, label = agaricus.train$label)
dtest <- xgb.DMatrix(agaricus.test$data, label = agaricus.test$label)
watchlist <- list(train = dtrain, eval = dtest)

## A simple xgb.train example:
param <- list(max_depth = 25,
              eta = 0.1,
              nthread = 1,
              gamma = 1,
              min_child_weight = 0,
              max_delta_step = 0,
              objective = "binary:logistic",
              eval_metric = "logloss",
              lambda = 1,
              subsample = 1,
              colsample_bytree = 1,
              colsample_bylevel = 1)

set.seed(0)
bst <- xgb.train(param, dtrain, nrounds = 5, watchlist)
preds1 <- predict(bst, dtest)

set.seed(1)
bst <- xgb.train(param, dtrain, nrounds = 5, watchlist)
preds2 <- predict(bst, dtest)

sum(preds1 == preds2) / 1611 # 1
```

Results:

```r
[1]	train-logloss:0.598865	eval-logloss:0.598766 
[2]	train-logloss:0.521571	eval-logloss:0.521380 
[3]	train-logloss:0.457066	eval-logloss:0.456791 
[4]	train-logloss:0.402509	eval-logloss:0.402156 
[5]	train-logloss:0.355884	eval-logloss:0.355457 
[6]	train-logloss:0.315707	eval-logloss:0.315210 
[7]	train-logloss:0.280851	eval-logloss:0.280288 
[8]	train-logloss:0.250443	eval-logloss:0.249816 
[9]	train-logloss:0.223792	eval-logloss:0.223104 
[10]	train-logloss:0.200342	eval-logloss:0.199595 
[11]	train-logloss:0.179639	eval-logloss:0.178836 
[12]	train-logloss:0.161310	eval-logloss:0.160451 
[13]	train-logloss:0.145042	eval-logloss:0.144129 
[14]	train-logloss:0.130541	eval-logloss:0.129580 
[15]	train-logloss:0.117620	eval-logloss:0.116612 
[16]	train-logloss:0.106089	eval-logloss:0.105034 
[17]	train-logloss:0.095783	eval-logloss:0.094682 
[18]	train-logloss:0.086561	eval-logloss:0.085413 
[19]	train-logloss:0.078325	eval-logloss:0.077127 
[20]	train-logloss:0.070728	eval-logloss:0.069617 
[21]	train-logloss:0.063902	eval-logloss:0.062874 
[22]	train-logloss:0.057765	eval-logloss:0.056816 
[23]	train-logloss:0.052243	eval-logloss:0.051369 
[24]	train-logloss:0.047292	eval-logloss:0.046484 
[25]	train-logloss:0.042811	eval-logloss:0.042070 
```

**Comment by Laurae2** [comment]
but when putting `lambda = 0`, predictions become different:

```r
> param <- list(max_depth = 5,
+               eta = 0.1,
+               nthread = 1,
+               gamma = 1,
+               min_child_weight = 0,
+               max_delta_step = 0,
+               objective = "binary:logistic",
+               eval_metric = "logloss",
+               lambda = 0,
+               subsample = 1,
+               colsample_bytree = 1,
+               colsample_bylevel = 1)
> set.seed(0)
> bst <- xgb.train(param, dtrain, nrounds = 25, watchlist)
[1]	train-logloss:0.625206	eval-logloss:0.625670 
[2]	train-logloss:0.576882	eval-logloss:0.576858 
[3]	train-logloss:0.514954	eval-logloss:0.515596 
[4]	train-logloss:0.476862	eval-logloss:0.476359 
[5]	train-logloss:0.440258	eval-logloss:0.439628 
[6]	train-logloss:0.416594	eval-logloss:0.415672 
[7]	train-logloss:0.389849	eval-logloss:0.389474 
[8]	train-logloss:0.372364	eval-logloss:0.371791 
[9]	train-logloss:0.352022	eval-logloss:0.352252 
[10]	train-logloss:0.338666	eval-logloss:0.340022 
[11]	train-logloss:0.319588	eval-logloss:0.321565 
[12]	train-logloss:0.293869	eval-logloss:0.296505 
[13]	train-logloss:0.284107	eval-logloss:0.287604 
[14]	train-logloss:0.274899	eval-logloss:0.278373 
[15]	train-logloss:0.255654	eval-logloss:0.260078 
[16]	train-logloss:0.234268	eval-logloss:0.239376 
[17]	train-logloss:0.223423	eval-logloss:0.228498 
[18]	train-logloss:0.208273	eval-logloss:0.213606 
[19]	train-logloss:0.193430	eval-logloss:0.197891 
[20]	train-logloss:0.188117	eval-logloss:0.192988 
[21]	train-logloss:0.181700	eval-logloss:0.186444 
[22]	train-logloss:0.167950	eval-logloss:0.173189 
[23]	train-logloss:0.158306	eval-logloss:0.164053 
[24]	train-logloss:0.152734	eval-logloss:0.158176 
[25]	train-logloss:0.142214	eval-logloss:0.148052 
> preds1 <- predict(bst, dtest)
> 
> set.seed(1)
> bst <- xgb.train(param, dtrain, nrounds = 25, watchlist)
[1]	train-logloss:0.632992	eval-logloss:0.632688 
[2]	train-logloss:0.584007	eval-logloss:0.582726 
[3]	train-logloss:0.544296	eval-logloss:0.542706 
[4]	train-logloss:0.511308	eval-logloss:0.509376 
[5]	train-logloss:0.464673	eval-logloss:0.463582 
[6]	train-logloss:0.432227	eval-logloss:0.431994 
[7]	train-logloss:0.412232	eval-logloss:0.411816 
[8]	train-logloss:0.372016	eval-logloss:0.373144 
[9]	train-logloss:0.339337	eval-logloss:0.341975 
[10]	train-logloss:0.315915	eval-logloss:0.319502 
[11]	train-logloss:0.304177	eval-logloss:0.307543 
[12]	train-logloss:0.279033	eval-logloss:0.283491 
[13]	train-logloss:0.257237	eval-logloss:0.262812 
[14]	train-logloss:0.234948	eval-logloss:0.241287 
[15]	train-logloss:0.218690	eval-logloss:0.225464 
[16]	train-logloss:0.210458	eval-logloss:0.217582 
[17]	train-logloss:0.196891	eval-logloss:0.204840 
[18]	train-logloss:0.186290	eval-logloss:0.194460 
[19]	train-logloss:0.181769	eval-logloss:0.190062 
[20]	train-logloss:0.170413	eval-logloss:0.179271 
[21]	train-logloss:0.170410	eval-logloss:0.179271 
[22]	train-logloss:0.161082	eval-logloss:0.167611 
[23]	train-logloss:0.151601	eval-logloss:0.158436 
[24]	train-logloss:0.145648	eval-logloss:0.153019 
[25]	train-logloss:0.137590	eval-logloss:0.145543 
> preds2 <- predict(bst, dtest)
> sum(preds1 == preds2) / 1611
[1] 0
```

It doesn't affect `alpha`.

**Comment by dataforager** [comment]
@Laurae2 Thanks a lot for your response. 

That's right. That's exactly what we see: lambda=0 results in different predicted probabilities where lambda=1 results in the same.

The lambda=1 behavior is what we would expect for runs with any value of lambda using different seeds (considering subsample=colsample_bylevel=colsample_bytree=1). There shouldn't be anything stochastic about these runs, right? (and so no sensitivity to seed) 

So, we're curious why lambda=0 results in different (using two different seeds) probabilities. It's just setting the regularization parameter to a different value. We're wondering why that would make a difference in the reproducibility of predicted probabilities in this 'deterministic' setting.

Thanks a lot for your help.

**Comment by khotilov** [comment]
You have a "perfect storm" situation here with both min_child_weight=0 and
lambda=0, which make it possible for a denominator (sum_hess + reg_lambda)
to disappear, and the learning becomes unstable. The problem goes away when
either of those parameters is set above zero. The results become dependent
on random seed most likely because some things start being sensitive to the
order of computations when the numerical instability comes into picture,
and we do still have the column order shuffling going on, even when all the
sampling parameters are set to 1.

It might make sense to set the lower allowed bound of min_child_weight to
something tiny above zero, like 1e-10.

**Comment by clinicalpersona** [comment]
It may also make sense to update the R help page. It now says this:

          2.2. Parameter for Linear Booster
            • ‘lambda’ L2 regularization term on weights. Default: 0
            • ‘lambda_bias’ L2 regularization term on bias. Default: 0
            • ‘alpha’ L1 regularization term on weights. (there is no
              L1 reg on bias because it is not important). Default: 0

which suggests that lambda, lambda_bias and alpha only apply to linear booster.
This is not true as @Laurae2 pointed out.

**Comment by Laurae2** [comment]
@clinicalpersona `lambda_bias` is a special case defined only for `gblinear`, so my initial answer was a bit wrong. But `alpha` and `lambda` are perfectly usable for `gbtree`.

I tend to refer to my own detailed table instead for xgboost parameters when I want to make sure an information is correct: https://sites.google.com/view/lauraepp/parameters - the official xgboost documentation has mistakes (sometimes misleading such as for the linear booster parameters).

As you can see on the picture below, the default is 1 for `gbtree` and 0 for `gblinear`. This can only be noticed on the source code of xgboost (or if you play a lot with the default parameters).

![image](https://user-images.githubusercontent.com/9083669/29965924-ec5afd38-8f0f-11e7-9a05-35fb4a4b2de9.png)

**Comment by clinicalpersona** [comment]
Great info, thank you for clarifying @Laurae2
